src/traits/fp.rs

```diff
@@ -58,3 +58,39 @@ pub trait FPMatrix {
     where
         F: Fn(Vec<f64>) -> f64;
 }
+
+/// Functional Programming for Complex Matrix
+pub trait FPComplexMatrix {
+    fn take_row(&self, n: usize) -> Self;
+    fn take_col(&self, n: usize) -> Self;
+    fn skip_row(&self, n: usize) -> Self;
+    fn skip_col(&self, n: usize) -> Self;
+    fn fmap<F>(&self, f: F) -> Self
+    where
+        F: Fn(Complex<f64>) -> Complex<f64>;
+    fn col_map<F>(&self, f: F) -> Self
+    where
+        F: Fn(Vec<Complex<f64>>) -> Vec<Complex<f64>>;
+    fn row_map<F>(&self, f: F) -> Self
+    where
+        F: Fn(Vec<Complex<f64>>) -> Vec<Complex<f64>>;
+    fn col_mut_map<F>(&mut self, f: F)
+    where
+        F: Fn(Vec<Complex<f64>>) -> Vec<Complex<f64>>;
+    fn row_mut_map<F>(&mut self, f: F)
+    where
+        F: Fn(Vec<Complex<f64>>) -> Vec<Complex<f64>>;
+    fn reduce<F, T>(&self, init: T, f: F) -> Complex<f64>
+    where
+        F: Fn(Complex<f64>, Complex<f64>) -> Complex<f64>,
+        T: Into<Complex<f64>>;
+    fn zip_with<F>(&self, f: F, other: &Self) -> Self
+    where
+        F: Fn(Complex<f64>, Complex<f64>) -> Complex<f64>;
+    fn col_reduce<F>(&self, f: F) -> Vec<Complex<f64>>
+    where
+        F: Fn(Vec<Complex<f64>>) -> Complex<f64>;
+    fn row_reduce<F>(&self, f: F) -> Vec<Complex<f64>>
+    where
+        F: Fn(Vec<Complex<f64>>) -> Complex<f64>;
+}
```

Comment:
Note: In next PRs, this can be removed and the handling added to `pub trait FPMatrix` (with a generic)

Reply:
Regarding the use of Complex in `src/traits/{fp, mutable}.rs`:

1. Let's avoid using `use num_complex::Complex` directly in these files, as Complex should only be available with the "complex" feature enabled.

2. You mentioned that a generic implementation is planned for the future. Given this, I agree it would be best to start with the generic implementation now rather than implementing Complex-specific versions that will need to be modified later.

Let's pivot to implementing these traits generically from the outset. This approach will:
- Provide a more flexible foundation
- Reduce duplicate code when supporting both real and complex numbers
- Align with your planned future direction

Could you share your thoughts on how you'd like to approach this generic implementation? We can discuss any specific challenges or design considerations you foresee.

Reply:
Done in PR: https://github.com/Axect/Peroxide/pull/71/commits/f2778957a20f24e9ef709482a4a7e0013c1b7788